VTUL/dlp-access
# Render collection organization tree for improved navigation on Collection Show pages

**Render collection organization tree for improved navigation on Collection Show pages.**
* * *

**JIRA Ticket**: (https://webapps.es.vt.edu/jira/browse/LIBTD-2291) (:star:)

* Other Relevant Links (Meeting note, project page, related pull requests, etc.)

# What does this Pull Request do? (:star:)
Render collection organization tree for improved navigation on Collection Show pages. This includes changes to the data model by adding a CollectionMap table that is linked to the Collection table. This stores the generated objects that represent each collection. This PR also adds a `heirarchy_path` field to the Collection table that represents each Collection ID that defines the parent-child relationships from the top-level collection down to the sub-collection being rendered.

# What's the changes? (:star:)
* Adds a CollectionMap table to data model to store the json object which represents the collection heirarchy.
* Links CollectionMap table entries to their appropriate collections
* Adds a `heirarchy_path` field to the collection table. This represents the parent-child relationship from the top-level collection down to the sub-collection being rendered. This field allows the collection heirarchy to "open" to the correct sub-collection and show it highlighted.
* Renders the collection heirarchy on the collection show page using the Material-UI tree-view plugin https://material-ui.com/components/tree-view/#customized-tree-view

# How should this be tested?
* This PR includes additions and changes to the data model. Because of this the preview generated by this PR will probably fail to deploy.
* The changes in this PR can be viewed here: https://libtd-2291.djf1n0m63xbku.amplifyapp.com/
* Go to a collection page and check that the hierarchy tree is rendering in the right column.
* Check that folders can be opened and closed
* Check that you can navigate to sub-collections by clicking on their entry in the hierarchy.
* Check that the correct collection or sub-collection is highlighted on each page 

# Additional Notes:
* branch: `LIBTD-2291`
* After changes are merged into dev the dev branch must be checked out locally and then `amplify push` must be run to push the data model changes up to the dev environment in Amplify.
* After the dev environment has finished updating, three lambda functions need to be run to populate the correct data into the dev environment tables.
* First run `collectionTitleEdit` https://console.aws.amazon.com/lambda/home?region=us-east-1#/functions/collectionTitleEdit This will modify the titles for IAWA sub-collections to NOT include identifiers.
* After `collectionTitleEdit` has completed running then `createCollectionMap` needs to be run. https://console.aws.amazon.com/lambda/home?region=us-east-1#/functions/createCollectionMap NOTE: THIS TABLE (Collectionmap) DOES NOT EXIST YET. It will be created when the dev environment is updated. After the table is created the table name needs to be added to the environment variables of this lambda function. Also, the ARN for the created table needs to be added to the collectionmap-policy for this lambda function. https://console.aws.amazon.com/iam/home#/roles/createCollectionMap-role-ld7k559y$jsonEditor?policyName=collectionmap-policy&section=permissions&step=edit Then the function can be run. I can help with this step if necessary. Just let me know.
* Finally, `createCollectionHeirarchy` needs to be run. https://console.aws.amazon.com/lambda/home?region=us-east-1#/functions/createCollectionHeirarchy

# Interested parties
@yinlinchen 

(:star:) Required fields
